Improve error message while setting error twice.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54464 Improve error message while setting error twice.**

In case where we accidentaly set an error twice on a Future, we get a
cryptic error like this:

```
Exception in thread pool task: !completed() INTERNAL ASSERT FAILED at "aten/src/ATen/core/ivalue_inl.h":534, please report a bug to PyTorch.
```

This PR, updates the error message to include some additional information about
what the previous error was.

Differential Revision: [D27249758](https://our.internmc.facebook.com/intern/diff/D27249758/)

pritamdamania committed:
Improve error message while setting error twice.

In case where we accidentaly set an error twice on a Future, we get a
cryptic error like this:

```
Exception in thread pool task: !completed() INTERNAL ASSERT FAILED at "aten/src/ATen/core/ivalue_inl.h":534, please report a bug to PyTorch.
```

This PR, updates the error message to include some additional information about
what the previous error was.

Differential Revision: [D27249758](https://our.internmc.facebook.com/intern/diff/D27249758/)

[ghstack-poisoned]

swolchok commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -2203,6 +2204,17 @@ TEST(FuturesTest, CollectAny) {
   ASSERT_EQ(c4->value().toInt(), 7);
 }
 
+TEST(FuturesTest, SetError) {
```
shouldn't this go in aten/src/ATen/test/ivalue_test.cpp?

swolchok commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -558,7 +558,11 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
-    AT_ASSERT(!completed());
+    TORCH_CHECK(
+        !eptr_,
+        "Error already set on the Future: ",
```
nit: s/the Future/this Future/ ?

swolchok reviewed (changes requested):
setErrorInternal looks fine to me, though I think @ezyang has more specific thoughts on error message wording than I do. please move the test

ezyang commented:
> though I think @ezyang has more specific thoughts on error message wording than I do

Wording seems fine, don't think this will actually show up for end users

pritamdamania committed:
Update on "Improve error message while setting error twice."


In case where we accidentaly set an error twice on a Future, we get a
cryptic error like this:

```
Exception in thread pool task: !completed() INTERNAL ASSERT FAILED at "aten/src/ATen/core/ivalue_inl.h":534, please report a bug to PyTorch.
```

This PR, updates the error message to include some additional information about
what the previous error was.

Differential Revision: [D27249758](https://our.internmc.facebook.com/intern/diff/D27249758/)

[ghstack-poisoned]

rohan-varma commented on aten/src/ATen/test/ivalue_test.cpp:
```diff
@@ -308,6 +309,18 @@ TEST(IValueTest, FutureExceptions) {
+    f1->setError(std::make_exception_ptr(std::runtime_error("bar")));
+    FAIL() << "Expected to throw";
+  } catch (std::exception& e) {
+    EXPECT_THAT(e.what(), ::testing::HasSubstr("Error already set"));
```
nit: also test that it has the original error message? 
EXPECT_THAT(e.what(), ::testing::HasSubstr("foo"));

pritamdamania committed:
Update on "Improve error message while setting error twice."


In case where we accidentaly set an error twice on a Future, we get a
cryptic error like this:

```
Exception in thread pool task: !completed() INTERNAL ASSERT FAILED at "aten/src/ATen/core/ivalue_inl.h":534, please report a bug to PyTorch.
```

This PR, updates the error message to include some additional information about
what the previous error was.

Differential Revision: [D27249758](https://our.internmc.facebook.com/intern/diff/D27249758/)

[ghstack-poisoned]

pritamdamania87 commented:
@swolchok Could you take another look at this PR? Thanks!

swolchok reviewed (approved):
Looks good, sorry for the delay!
